Introducing ScalarTypeType and LayoutType

charliechen0401 opened the pull request:
Summary:
As the title,
It's step 1 in https://github.com/pytorch/pytorch/pull/30694#issuecomment-564205276.

Not using those types in any other place.

Test Plan: Making sure all unit tests and build pass successfully.

Differential Revision: D18916246

charliechen0401 reviewed (commented):
Probably some issue related to permission which doesn't allow me to specify reviewers to this PR. :(

smessmer commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1708,4 +1710,55 @@ struct CAFFE2_API InterfaceType : public NamedType {
 
+template <TypeKind K>
+struct EnumerationType : public Type {
+  static const TypeKind Kind = K;
```
nit: can be constexpr

smessmer commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1708,4 +1710,55 @@ struct CAFFE2_API InterfaceType : public NamedType {
+struct CAFFE2_API LayoutType : public EnumerationType<TypeKind::LayoutType> {
+static LayoutTypePtr create() {
+  return LayoutTypePtr(
+      new LayoutType()); // NOLINT(modernize-make-shared)
```
nit: use `make_shared`

charliechen0401 replied:
I will do this in a separate diff and fix it for all types listed here.

smessmer commented:
Please also wait for @suo to take a look before landing.

charliechen0401 commented:
Thank you @smessmer. I will wait for comments from @suo as well.
Also I can tag reviewers now. Thanks for the steps to help me join pytorch org :)

Commit:
Introducing ScalarTypeType and LayoutType (#31074)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/31074

As the title,
It's step 1 in https://github.com/pytorch/pytorch/pull/30694#issuecomment-564205276.

Not using those types in any other place.

Test Plan: Making sure all unit tests and build pass successfully.

Differential Revision: D18916246

fbshipit-source-id: 3f5c74967eedfb0a31d2f983e9df6b47fd1b9e52

kostmo commented:
## CircleCI build failures summary
As of commit f336f7e:
* **2/2** broken upstream at merge base c6a8f88 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=c6a8f884d83e1904b6cbcaf7b924a36888330842&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f336f7e35ed60ff8e13e25608cd166bbe633d3f4).

### 2 upstream failures recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [caffe2_onnx_py3_6_clang7_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75443288)
* [caffe2_onnx_py2_gcc5_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75445441)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

suo commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1710,4 +1712,54 @@ struct CAFFE2_API InterfaceType : public NamedType {
+    new ScalarTypeType()); // NOLINT(modernize-make-shared)
+}
+std::string str() const override {
+return "ScalarType";
```
We should override `python_str()` to report "dtype"—unfortunately, aten/c10 call this concept ScalarType, but the PyTorch frontend calls it "dtype". For users it will be confusing because they don't have any notion of what "ScalarType" is.

suo commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -1710,4 +1712,54 @@ struct CAFFE2_API InterfaceType : public NamedType {
 };
 
+template <TypeKind K>
+struct EnumerationType : public Type {
```
It is hard to review this PR without the other context surrounding it. For example, currently `EnumerationType` provides no value since it does basically nothing. I'm sure at some point there will be common methods that are useful to define for enumerable types, but at this point I don't know what they are.

suo commented on aten/src/ATen/core/type.cpp:
```diff
@@ -116,6 +116,14 @@ DeviceObjTypePtr DeviceObjType::get() {
   static auto value = DeviceObjType::create();
   return value;
 }
+ScalarTypeTypePtr ScalarTypeType::get() {
```
Please clang-format the region of the code that you changed :)

suo commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -354,6 +354,9 @@ inline IValue toIValue(
     case TypeKind::FloatType:
       return py::cast<double>(obj);
     case TypeKind::IntType:
+    // TODO(xintchen): Handling LayoutType and ScalarTypeType correctly.
```
Once again, it's hard to know why this change is in here, because as-is this PR doesn't ever use LayoutType or ScalarTypeType anywhere. As such, it's not self-contained enough to review properly.

suo reviewed (commented):
@charliechen0401 I would have appreciated if you had waited for me to take a look at this PR before landing it. As is, this PR is not self-contained enough to review it well, so it would have been nice to at least see the upper parts of the PR stack before reviewing/landing this code.

Please see inline comments for some changes we should make before continuing.

suo commented:
If I'm taking too long to get around to reviewing this, feel free to ping me directly on messenger for a review next time.
